waitress/server.py

```diff
@@ -359,5 +373,8 @@ def getsockname(self):
         def fix_addr(self, addr):
             return ('localhost', None)
 
+        def get_server_name(self, ip):
+            return 'localhost'
```

Comment:
I'm not sure that `'localhost'` is the Right Thing(TM) for Unix-domain sockets. `None` seems more appropriate.

Reply:
@tseaver this is what is placed in `SERVER_NAME` in the WSGI environ. `None` IMHO is less valid than `localhost`.

Reply:
from PEP0333:

> `SERVER_NAME` and `SERVER_PORT` can never be empty strings, and so are always required.

Reply:
Do note that previously `SERVER_NAME` would have been set to `unix` due to `getsockname()` returning `unix` for unix sockets.

Reply:
`unix` seems more correct than `localhost`:  that name is only relevant for IP sockets.

Reply:
Not if the `SERVER_NAME` is used for generation of URL's (which it is used to do as a fallback if `HTTP_HOST` is not set). `localhost` has a defined location, and should always resolve to 127.0.0.1, `unix` depending on DNS configuration may lead to random URL's being generated/used by clients redirecting them to a system named `unix`.

Reply:
> When combined with SCRIPT_NAME and PATH_INFO, these variables can be used to complete the URL. Note, however, that HTTP_HOST, if present, should be used in preference to SERVER_NAME for reconstructing the request URL. See the URL Reconstruction section below for more detail. SERVER_NAME and SERVER_PORT can never be empty strings, and so are always required.

From PEP 0333. Using `unix` as the `SERVER_NAME` is far worse than using `localhost`.

Reply:
It is actually weird to be trying to do HTTP/WSGI over Unix-domain sockets.

Presumably, there is an HTTP proxy on the other side:  wouldn't we be better off using whatever *its* values are for `SERVER_NAME` / `SERVER_PORT`, since the WSGI app  might be using them to generate URLs?  I think URL generation is why PEP 333 requires them.

Reply:
> From PEP 0333. Using unix as the SERVER_NAME is far worse than using localhost.

I don't see how any URL generation using `localhost` and whatever port number we are getting from `None` can *possibly* yield valid URLs.

Reply:
`SERVER_NAME` is always set locally, `HTTP_HOST` can be passed by nginx or some proxy, and should be set correctly.

`SERVER_PORT` right now is always the `effective_port` which is set to `/tmp/waitress.test.sock` or wherever the socket resides in the case of the unix socket... and is not correct. We also don't respect `X-Forwarded-Port` or anything else to allow that to be overridden.

https://github.com/Pylons/waitress/blob/4ce8f82082a76aa62148b6f3300d16ec4184202e/waitress/task.py#L547

There was even discussion about it on Web-SIG: https://mail.python.org/pipermail/web-sig/2012-January/005097.html

Right now gunicorn for example just appends an empty string for `SERVER_PORT` after trying to parse the `HTTP_HOST` header: https://github.com/benoitc/gunicorn/blob/91974f0f44080fdd6f2885832d101474671c4b9b/gunicorn/http/wsgi.py#L175

And `SERVER_NAME` gets set to the path of the unix domain socket if the proxy doesn't send a `Host` header...

This is a new issue though, and we can open one to fix this, but using `unix` as the `SERVER_NAME` is IMHO, wrong.

Reply:
I have opened issue https://github.com/Pylons/waitress/issues/207 for this.

Reply:
@tseaver

> I don't see how any URL generation using localhost and whatever port number we are getting from None can possibly yield valid URLs.

Right now we would generate it using `localhost` as the `SERVER_NAME` and `/tmp/waitress.test-43692.sock` as `SERVER_PORT` which would yield the following URL's:

`http://localhost:/tmp/waitress.test-43692.sock/other/path/info`.

Which is a "valid" URL.

I'd much rather we create URL's using "localhost" than "unix". "localhost" is guaranteed to resolve to 127.0.0.1, `unix` may resolve to whatever happens to be out there on the internet.

If the URL is returned to a user with `unix` set in the URL, and they try to connect to it, they will do a DNS request for `unix` followed by `unix.<whatever search path they have setup for their system's DNS>`.

From my logs at home while trying to run a curl `http://unix/`.

```
Sep  6 19:43:23 Breached unbound: [30770:0] info: 10.10.10.130 unix.home.arpa. A IN NXDOMAIN 0.000000 1 82
Sep  6 19:43:23 Breached unbound: [30770:0] info: 10.10.10.130 unix. A IN NXDOMAIN 0.000000 1 97
```

For example:

```
Riley:~ xistence$ dig unix

; <<>> DiG 9.10.6 <<>> unix
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NXDOMAIN, id: 43892
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 1, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;unix.				IN	A

;; AUTHORITY SECTION:
.			3600	IN	SOA	a.root-servers.net. nstld.verisign-grs.com. 2018090601 1800 900 604800 86400

;; Query time: 96 msec
;; SERVER: 10.10.10.11#53(10.10.10.11)
;; WHEN: Thu Sep 06 13:43:15 MDT 2018
;; MSG SIZE  rcvd: 108
```

Went all the way to the root DNS servers just to get a negative response.

Whereas gets resolved locally:

```
Riley:~ xistence$ dig localhost

; <<>> DiG 9.10.6 <<>> localhost
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 29739
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;localhost.			IN	A

;; ANSWER SECTION:
localhost.		10800	IN	A	127.0.0.1

;; Query time: 66 msec
;; SERVER: 10.10.10.11#53(10.10.10.11)
;; WHEN: Thu Sep 06 13:43:29 MDT 2018
;; MSG SIZE  rcvd: 54
```

This just means potential clients aren't connecting to the internet/asking for information that is invalid.

I am not suggesting it's the best behaviour, but I do think it is better than arbitrarily using `unix` for `SERVER_NAME`. As I mentioned, I've created/opened a new ticket to fix the behaviour as best as possible by using information provided by the proxy server that sits in front of the unix socket, so that the information is correct.